File: solution/826. Most Profit Assigning Work/826.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        const int n = profit.size();
        
        vector<pair<int, int>> A(n);
        for (int i = 0; i < n; ++i)
            A[i] = {difficulty[i], profit[i]};
        
        sort(A.begin(), A.end());
        
        sort(worker.begin(), worker.end());
        
        int res = 0, i = 0, mx = 0;
        priority_queue<int> pq;
        for (int w : worker) {
            while (i < n && A[i].first <= w)
                mx = max(mx, A[i++].second);
            res += mx;
        }
        return res;
    }
};

File: solution/826. Most Profit Assigning Work/826.py
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        A = sorted(zip(difficulty, profit))
        res = i = mx = 0
        for w in sorted(worker):
            while i < len(A) and A[i][0] <= w:
                mx = max(mx, A[i][1])
                i += 1
            res += mx
        return res
    
